pvlib/singlediode.py

```diff
@@ -679,6 +682,105 @@ def _lambertw(photocurrent, saturation_current, resistance_series,
     return out
 
 
+def _w_psi(i, il, io, gsh, a):
+    ''' Computes W(psi), where psi = io * rsh / a exp((il + io - i) * rsh / a)
+    This term is part of the equation V=V(I) solving the single diode equation
+    V = (il + io - i)*rsh - i*rs - a W(psi)
+
+    Parameters
+    ----------
+    i : numeric
+        Current (A)
+    il : numeric
+        Photocurrent (A)
+    io : numeric
+        Saturation current (A)
+    gsh : numeric
+        Shunt conductance (1/Ohm)
+    a : numeric
+        The product n*Ns*Vth (V).
+
+    Returns
+    -------
+    lambertwterm : numeric
+        The value of W(psi)
+
+    '''
+    with np.errstate(over='ignore'):
+        argW = (io / (gsh * a) *
+                np.exp((-i + il + io) /
+                       (gsh * a)))
+    lambertwterm = np.array(lambertw(argW).real)
+
+    idx_inf = np.logical_not(np.isfinite(lambertwterm))
+    if np.any(idx_inf):
+        # Calculate using log(argW) in case argW is really big
+        logargW = (np.log(io) - np.log(gsh) -
+                   np.log(a) +
+                   (-i + il + io) /
+                   (gsh * a))[idx_inf]
+
+        # Six iterations of Newton-Raphson method to solve
+        #  w+log(w)=logargW. The initial guess is w=logargW. Where direct
+        #  evaluation (above) results in NaN from overflow, 3 iterations
+        #  of Newton's method gives approximately 8 digits of precision.
+        w = logargW
+        for _ in range(0, 6):
+            w = w * (1. - np.log(w) + logargW) / (1. + w)
+        lambertwterm[idx_inf] = w
+    return lambertwterm
+
+
+def _imp_est(i, il, io, rs, gsh, a):
+    wma = _w_psi(i, il, io, gsh, a)
+    f = (il + io - i) / gsh - i * rs - a * wma
+    fprime = -rs - 1. / (gsh * (1. + wma))
+    return -f / fprime
+
+
+def _split_on_gsh(gsh):
+    # Determine indices where 0 < Gsh requires implicit model solution
+    idx_p = 0. < gsh
+    # Determine indices where 0 = Gsh allows explicit model solution
+    idx_z = 0. == gsh
+    return idx_p, idx_z
+
+
+def _imp_zero(i, il, io, rs, gsh, a):
+    ''' Root of this function is at imp
+    '''
+    idx_p, idx_z = _split_on_gsh(gsh)
+    res = np.full_like(i, np.nan, dtype=np.float64)
+
+    if np.any(idx_z):
+        # explicit solution for gsh=0
+        t = il[idx_z] + io[idx_z] - i[idx_z]
+        res[idx_z] = i[idx_z] / t - np.log(t / io[idx_z])
+    if np.any(idx_p):
+        res[idx_p] = _imp_est(i[idx_p], il[idx_p], io[idx_p], rs[idx_p],
+                              gsh[idx_p], a[idx_p]) - i[idx_p]
+    return res
+
+
+def _imp_zero_prime(i, il, io, rs, gsh, a):
+    ''' Derivative of _imp_zero with respect to current i
+    '''
+    idx_p, idx_z = _split_on_gsh(gsh)
+    res = np.full_like(i, np.nan, dtype=np.float64)
+    if np.any(idx_z):
+        # explicit solution for gsh=0
+        t = il[idx_z] + io[idx_z] - i[idx_z]
+        res[idx_z] = 2. / t + i[idx_z] / t**2.
```

Comment:
Depending on whether the previous comment is in error, this may also need an additional term